### PR TITLE
Rating, RatingStar

### PR DESCRIPTION
**Pull request by daveols**
Introducing the Rating component, a cheeky lil extra for dessert.

🙏

### CODE REVIEW

**Comment by daveols** [comment]
@jxom the theming is incomplete -- would like your feedback and guidance before implementing as per existing standards. Cheers.

File: src/Rating/Rating.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+type State = {
+  rating: ?number,
+  isSelecting: boolean,
+  isStateDirty: boolean,
+  selectedIndex: ?number
+};
+
+class Rating extends Component<Props, State> {
+  static defaultProps = {
+    className: null,
```
Comment (jxom) [inline src/Rating/Rating.js]: Lets make the `null`ed default props actually `undefined`, so they conform with the Flow types (I just figured out I was also doing this wrong the whole time too). `className?: string` is valid when `className` is a `string` or `undefined` whereas `className: ?string` is valid when `className` is `string` or `null`.

File: src/Rating/Rating.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import times from 'lodash/times';
+
+import type { Size } from '../types';
+import RatingStar from './RatingStar';
+import { Rating as _Rating } from './styled';
+
+type Props = {
+  className?: string,
+  size: Size,
+  rating?: number,
```
Comment (jxom) [inline src/Rating/Rating.js]: This prop should be something like `defaultRating`? I'd expect `rating` to be static in nature (like `defaultValue` and `value` with input fields), meaning that the Rating component state becomes controlled by the user with `onRate` and `rating`. Maybe we could have `rating` and `defaultRating` coexist, and the user can chose if they want to control it or not?

File: src/Rating/Rating.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  };
+
+  state = {
+    rating: this.props.rating,
+    isSelecting: false,
+    isStateDirty: false,
+    selectedIndex: null
+  };
+
+  static getDerivedStateFromProps = (nextProps: Props, prevState: State) => {
```
Comment (jxom) [inline src/Rating/Rating.js]: Is there a reason to use `gDSFP` over `cDU`?

File: src/Rating/Rating.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import RatingStar from './RatingStar';
+import { Rating as _Rating } from './styled';
+
+type Props = {
+  className?: string,
+  size: Size,
+  rating?: number,
+  maxRating?: string,
+  onRate?: Function,
+  disabled?: boolean
```
Comment (jxom) [inline src/Rating/Rating.js:15]: Can we add some more effective styling to disabled stars (like grayed out)? They look the same as active stars atm.

File: src/Rating/Rating.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  isStateDirty: boolean,
+  selectedIndex: ?number
+};
+
+class Rating extends Component<Props, State> {
+  static defaultProps = {
+    className: null,
+    size: 'regular',
+    rating: 0,
+    maxRating: 5,
```
Comment (jxom) [inline src/Rating/Rating.js:29]: Can we add an example in the docs on different `maxRating`s?

File: src/Rating/Rating.mdx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+## Basic Usage
+
+The rating callback returns `{ rating, maxRating }`.
+
+<Playground>
+  <Rating  onRate={({ rating }) => console.log(rating)}/>
+</Playground>
+
+## Controlled
```
Comment (jxom) [inline src/Rating/Rating.mdx]: This title is misleading. To me, "controlled" means that the state will be controlled by the user, and not `Rating` itself. Maybe **Default values** is more appropriate.

File: src/Rating/styled.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+export const Rating = styled.div``;
+
+const starSizes = {
+  small: css`
+    & {
+      width: 1rem;
+      height: 1rem;
+    }
+    & {
+      ${theme('rating.sizes.small')};
```
Comment (jxom) [inline src/Rating/styled.js]: `rating` -> `Rating`

File: src/Rating/styled.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  height: 1.5rem;
+
+  & {
+    ${theme('rating.Star.base')};
+  }
+
+  ${props => starSizes[props.size]} /* Extend size styles */;
+
+  & path {
+    fill: ${props => (props.active ? theme('rating.Star.color') || palette('black') : palette('whiteDarkest'))};
```
Comment (jxom) [inline src/Rating/styled.js]: I reckon we should default the theme color to gold

**Review by jxom — CHANGES_REQUESTED** [review]

File: src/Rating/Rating.js (hunk 2 above)
Comment (daveols) [reply src/Rating/Rating.js]: I'm going to leave this as-is for now. If we find people need more explicit control we can always add it! Will try to explain more clearly how it works in the docs.

File: src/Rating/Rating.js (hunk 3 above)
Comment (daveols) [reply src/Rating/Rating.js]: I like it ! ! ! ! !

I'll refactor for you tho Jakey boi 😄

Comment (daveols) [reply src/Rating/Rating.js]: Upon further review it does make sense to leave it in GDSFP. The internal state needs to be conditionally updated based on prop change.

File: src/Rating/Rating.js (hunk 4 above)
Comment (daveols) [reply src/Rating/Rating.js:15]: I think it makes sense for them to look the same. The case where someone might want to show an existing rating which is not editable is likely more common than an updatable rating which is "unlocked" by certain actions like how disabled behaves with buttons.

File: src/Rating/Rating.mdx (hunk 6 above)
Comment (daveols) [reply src/Rating/Rating.mdx]: It's not exactly controlled in the traditional sense. I'll update the docs to be more descriptive.

File: src/Rating/Rating.js (hunk 2 above)
Comment (daveols) [reply src/Rating/Rating.js]: Changed my mind again. Going to do defaultRating as per https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

**Pull request merged**